nexus/db-queries/src/db/queries/next_item.rs

```diff
@@ -584,13 +603,328 @@ impl<Item> ShiftGenerator<Item> for DefaultShiftGenerator<Item> {
     }
 }
 
+/// Select the next available item from a table, using a self-join.
+///
+/// This query is similar in spirit to the above `NextItem` query. However, it's
+/// implementation is different, and specifically designed to limit memory
+/// consumption of the overall query.
+///
+/// CockroachDB eagerly evaluates subqueries. This leads to high memory usage
+/// for the basic `NextItem` query, since that relies on a subquery that uses
+/// `generate_series()` to create the list of all possible items. That entire
+/// list must be materialized and held in memory.
+///
+/// In contrast, this query is implemented using a self-join between the
+/// existing table to be searched and the "next entry" (e.g., the value of the
+/// target column plus 1). This can result in lower memory consumption, since
+/// that series need not be stored in memory. Note that this relies on an index
+/// on the item column, which may be partial.
+///
+/// The full query looks like this:
+///
+/// ```sql
+/// SELECT IF(
+///     -- This condition detects if the table is empty.
+///     EXISTS(
+///         SELECT
+///             1
+///         FROM
+///             <table>
+///         WHERE
+///             time_deleted IS NULL
+///         LIMIT 1
+///     ),
+///     -- If the table is _not_ empty, we do the self-join between `item` and
+///     -- `item + 1`, and take the first value where `item` is NULL, i.e., the
+///     -- smallest value of `item + 1` where there is no corresponding `item`
+///     -- in the table.
+///     (
+///         SELECT next_item AS <item_column>
+///         FROM (
+///             SELECT
+///                 <item_column> + 1 AS next_item
+///             FROM
+///                 <table>
+///             WHERE
+///                 <scope_column> = <scope_key> AND
+///                 time_deleted IS NULL
+///         )
+///         LEFT OUTER JOIN
+///             <table>
+///         ON
+///             (<scope_column>, next_item <= <max_item>, time_deleted IS NULL) =
+///             (<scope_key>, TRUE, TRUE)
+///         WHERE
+///             <item_column> IS NULL AND next_item <= <max_item>
+///         ORDER BY next_item
```

Comment:
Does the `ORDER BY next_item` matter here? We can keep it in for tests determinism, but it basically means "we're picking the lowest possible free item, immediately after all existing items", right?

It's not randomly distributed, but omitting this would mean "we could pick *any* next lowest possible free item after all existing items", right?

Reply:
The ordering was intentional, yeah. It doesn't quite mean "after all existing items", it really means "the lowest possible free item". Those differ if there are gaps, for example. We fill the gaps first, with this query.

I think that's usually what we want to do. The only place we don't already do that is when we allocate a VNI for a new VPC (which this PR doesn't change) and when we select a MAC for a NIC (which this PR _does_ change). I don't have a strong opinion about whether it _matters_, but I do like the determinism for most situations.